Odroid h4 rel v0.9.1

miczyg1 opened the pull request:
Links to test results are in https://github.com/Dasharo/open-source-firmware-validation/pull/881#issuecomment-2962937001.

*ref: sha-1332*

miczyg1 committed:
mb/hardkernel/odroid-h4: Add Net Card 2 support

Add necessary code to support 4x1 bifurcation in M.2 NVMe slot
for Net Card 2. Create a new build variant supporting Net Card 2.

TEST=Boot Linux on ODROID H4 Ultra with Net Card 2 and run iperf
in all 4 ports from the Net Card 2.

Upstream-Status: Pending
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 committed:
configs/config.hardkernel_odroid_h4*: Enable quiet and fast boot

Change-Id: Ic6fdf94450ce5ed5884a7851e53e68c43ecdf6b7
Upstream-Status: Inappropriate [custom feature]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
vendorcode/dasharo,soc/intel: Add IBECC option

Add code to support new menu option for In-band ECC feature
control.

Upstream-Status: Inappropriate [custom feature]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mainboard/hardkernel/odroid-h4: Add IBECC configuration

Add In-band ECC configuration in devicetree necessary to fill
FSP UPDs to enable IBECC.

Upstream-Status: Inappropriate [custom feature]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
payloads/external/edk2: Add IBECC option support

Add code to enable In-band ECC feature option in setup menu.

Change-Id: I1acdf89ccf5e390d058e6d2a3893c1545282f5fc
Upstream-Status: Inappropriate [custom feature]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
payloads/external/edk2: Update revision for new memory menu PCDs

Paylaod revision has to be updated with introduction of new PCD.
New PCD controls the visibility of XMP profile selection so that
it is independent of IBECC option.

Upstream-Status: Inappropriate [custom configs]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mb/hardkernel/odroid-h4/data.vbt: Update VBT for newest FSP GOP

Fixes the issue with display not working in the firmware.
New FSP updated VBT definitions. Old VBTs no longer work with new FSP.
Update VBT so that it works with up to date ADL-N FSP.

TEST=Boot ODROID H4 Ultra with monitor plugged into all DP and HDMI
ports and observe boot logo on the screen.

Upstream-Status: Pending
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mb/hardkernel/odroid-h4/devicetree.cb: Disable DSP

DSP is not needed for audio to work. Moreover, it causes problems
with driver binding in Windows as it changes how the HDA controller
is seen in Windows.

Upstream-Status: Pending
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.hardkernel_odroid_h4*: Enable FW performance measurement

FW performance measurement is required to measure difference in boot
time with fast boot.

Change-Id: I28071058985f16e4cc1cee936c7a380862e5be64
Upstream-Status: Inappropriate [custom configs]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
.github/workflows/build.yml: Build netcard variant of ODROID

Add build checks for new ODROID build variant.

Change-Id: Icf2f3645e295b721356252e996900fd14f278b3b
Upstream-Status: Inappropriate [CI configuration]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.hardkernel_odroid_h4*: Bump to v0.9.1-rc1

Change-Id: I7b695dc1e334bea77480e58a6e1695f440eed0a5
Upstream-Status: Inappropriate [custom configs]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mb/hardkernel/odroid-h4/gpio.c: Fix GPP_A8 macro

Change-Id: Id66590bbb4aba3f19d448ebcb08aa544f25b975c
Upstream-Status: Pending
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
payloads/external/iPXE: Update revision to 2025.03

Update iPXE revision so that coreboot can be built with newer
toolchains.

Upstream-Status: Inappropriate [custom configs]
Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented:
All features implemented here were verified in automated tests in separate PRs.

I just need to rebase it, because it got outdated (mainly configs and EDK2 revision in them)

SergiiDmytruk commented on src/mainboard/hardkernel/odroid-h4/Kconfig:
```diff
@@ -80,4 +80,21 @@ config ODROID_H4_ENABLE_SAGV
+	bool "Enable 4x1 netcard support in M.2 NVME slot"
+	default n
+	help
+	  Enables 4x1 PCIe netcard support in the project. Enable this
```
nit: should it be "Net Card" instead of "netcard"?  That seems to be the name at https://www.hardkernel.com/shop/h3-h2-net-card-2/

miczyg1 replied:
Changed: https://github.com/Dasharo/coreboot/pull/694/commits/14a3fe7f4160428d51f6c5a84047b7fcc7fa0121#diff-813d3d376d7b2d2b25b6ec1a4862a97ca297b11cba73b9ce5b5a33e1784664c9R84

SergiiDmytruk commented on src/mainboard/hardkernel/odroid-h4/gpio.c:
```diff
@@ -273,8 +278,14 @@ static const struct pad_config gpio_table[] = {
 	PAD_NC(GPP_D4, NONE),
+#if CONFIG(ODROID_H4_NETCARD_SUPPORT)
+	/* GPP_D5 - GPIO */
+	/* Do not use SRCCLKREQ0# to PCIE RP 9 (NVMe), as the clock becomes shared  */
```
nit:
```suggestion
	/* Do not use SRCCLKREQ0# to PCIE RP 9 (NVMe), as the clock becomes shared */
```

miczyg1 replied:
Fixed: https://github.com/Dasharo/coreboot/pull/694/commits/14a3fe7f4160428d51f6c5a84047b7fcc7fa0121#diff-723e4fc0661623b211e2c53c59c46d8610e8ac9b9a91bc08adc4d92ab5673473R283

SergiiDmytruk commented on src/mainboard/hardkernel/odroid-h4/overridetree_netcard.cb:
```diff
@@ -0,0 +1,39 @@
+				.PcieRpL1Substates = L1_SS_DISABLED,
+			}"
+
+			smbios_slot_desc	"SlotTypeM2Socket3" "SlotLengthOther"
+						"M.2/M 2280 (M2_SSD1)" "SlotDataBusWidth4X"
```
Same SMBIOS data regardless whether it's 1x4 or 4x1?

miczyg1 replied:
@SergiiDmytruk SMBIOS specification is bifurcation agnostic. It only describes how many physical lanes are in the slot. SLot always will have 4 lanes, because that is how hardware is designed.

SergiiDmytruk commented on configs/config.msi_ms7e06_ddr5:
```diff
@@ -65,6 +65,7 @@ CONFIG_EDK2_DASHARO_CHIPSET_CONFIG=y
 CONFIG_EDK2_DASHARO_POWER_CONFIG=y
 CONFIG_EDK2_DASHARO_PCI_CONFIG=y
 CONFIG_EDK2_DASHARO_MEMORY_CONFIG=y
+CONFIG_EDK2_DASHARO_SPD_PROFILE_OPTION=y
```
`config.emulation_qemu_x86_q35_uefi_all_menus` also needs this line.

miczyg1 replied:
As EDK2 revision has been moved to Kconfig file in edk2 directory, the chanegs in configs have been dropped.

SergiiDmytruk reviewed (commented):
> I just need to rebase it

And add `Upstream-Status:` to commits.

https://github.com/Dasharo/coreboot/pull/694/commits/28a562f607ff20ae32f9c5b503e125740f268d42 could explain the removal.

miczyg1 commented:
> > I just need to rebase it
> 
> And add `Upstream-Status:` to commits.
> 
> [28a562f](https://github.com/Dasharo/coreboot/commit/28a562f607ff20ae32f9c5b503e125740f268d42) could explain the removal.

Yes, cleaned up all commits, added Upstream-Status and elaborated commit messages.

SergiiDmytruk commented:
I've aligned 2 configs for Odroid (including adding missing `CONFIG_DASHARO=y` to the new one), squashed 1 commit and fixed a couple typos.  Changes:

<details><summary>git range-diff</summary>
<p>

```
 1:  14a3fe7f416 !  1:  f954aa9eac6 mb/hardkernel/odroid-h4: Add Net Card 2 support
    @@ Commit message

         Upstream-Status: Pending
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
    +    Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

      ## configs/config.hardkernel_odroid_h4_netcard (new) ##
     @@
    @@ configs/config.hardkernel_odroid_h4_netcard (new)
     +CONFIG_PCIEXP_DEFAULT_MAX_RESIZABLE_BAR_BITS=37
     +CONFIG_ODROID_H4_NETCARD_SUPPORT=y
     +CONFIG_HAVE_IFD_BIN=y
    -+CONFIG_EDK2_BOOTSPLASH_FILE="3rdparty/dasharo-blobs/dasharo/bootsplash.bmp"
     +CONFIG_TPM_MEASURED_BOOT=y
    ++CONFIG_EDK2_BOOTSPLASH_FILE="3rdparty/dasharo-blobs/dasharo/bootsplash.bmp"
     +CONFIG_POWER_STATE_OFF_AFTER_FAILURE=y
     +CONFIG_HAVE_ME_BIN=y
     +CONFIG_DASHARO_PREFER_S3_SLEEP=y
    @@ configs/config.hardkernel_odroid_h4_netcard (new)
     +CONFIG_DRIVERS_GENERIC_CBFS_UUID=y
     +# CONFIG_DRIVERS_MTK_WIFI is not set
     +CONFIG_TPM2=y
    ++CONFIG_INTEL_TXT_LOGGING=y
    ++CONFIG_INTEL_CBNT_SUPPORT=y
    ++CONFIG_INTEL_CBNT_LOGGING=y
    ++# CONFIG_INTEL_CBNT_GENERATE_KM is not set
    ++# CONFIG_INTEL_CBNT_GENERATE_BPM is not set
    ++CONFIG_INTEL_CBNT_KEY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/km_sample.bin"
    ++CONFIG_INTEL_CBNT_BOOT_POLICY_MANIFEST_BINARY="3rdparty/dasharo-blobs/cbnt/bpm_sample.bin"
     +CONFIG_BOOTMEDIA_LOCK_CONTROLLER=y
     +CONFIG_BOOTMEDIA_LOCK_WPRO_VBOOT_RO=y
     +CONFIG_BOOTMEDIA_LOCK_IN_VERSTAGE=y
    @@ configs/config.hardkernel_odroid_h4_netcard (new)
     +CONFIG_IPXE_ADD_SCRIPT=y
     +CONFIG_IPXE_SCRIPT="3rdparty/dasharo-blobs/dasharo/dasharo.ipxe"
     +CONFIG_IPXE_CUSTOM_BUILD_ID="0123456789"
    ++CONFIG_DASHARO=y
     +CONFIG_EDK2_SKIP_PS2_DETECT=y
     +CONFIG_EDK2_ENABLE_IPXE=y
     +# CONFIG_EDK2_SECURE_BOOT_DEFAULT_ENABLE is not set
    @@ configs/config.hardkernel_odroid_h4_netcard (new)
     +CONFIG_EDK2_SETUP_PASSWORD=y
     +CONFIG_EDK2_DASHARO_SYSTEM_FEATURES=y
     +CONFIG_EDK2_DASHARO_SECURITY_OPTIONS=y
    ++CONFIG_EDK2_DASHARO_INTEL_ME_OPTIONS=y
    ++CONFIG_EDK2_DASHARO_INTEL_ME_CBNT_STATUS=y
     +CONFIG_EDK2_DASHARO_USB_CONFIG=y
     +CONFIG_EDK2_DASHARO_NETWORK_CONFIG=y
     +CONFIG_EDK2_DASHARO_POWER_CONFIG=y
 2:  b95db07584a =  2:  03090dfea27 mb/hardkernel/odroid-h4/data.vbt: Update VBT for newest FSP GOP
 3:  62a38559ad1 !  3:  f5e0c789216 configs/config.hardkernel_odroid_h4: Enable quiet and fast boot
    @@ Metadata
     Author: Michał Żygowski <michal.zygowski@3mdeb.com>

      ## Commit message ##
    -    configs/config.hardkernel_odroid_h4: Enable quiet and fast boot
    +    configs/config.hardkernel_odroid_h4*: Enable quiet and fast boot

    +    Change-Id: Ic6fdf94450ce5ed5884a7851e53e68c43ecdf6b7
         Upstream-Status: Inappropriate [custom feature]
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

 4:  9275e59c89e <  -:  ----------- configs/config.hardkernel_odroid_h4: Enable ME configuration menu
 5:  a0f0c3b4387 =  4:  73b76ffc230 mb/hardkernel/odroid-h4/devicetree.cb: Disable DSP
 6:  c7a7356a2d0 =  5:  2f0a47bdea5 vendorcode/dasharo,soc/intel: Add IBECC option
 7:  4df493135d2 =  6:  8fdb794d9ac mainboard/hardkernel/odroid-h4: Add IBECC configuration
 8:  d99ee2793b3 !  7:  321a6cecc7a payloads/external/edk2: Add IBECC option support
    @@ Metadata
      ## Commit message ##
         payloads/external/edk2: Add IBECC option support

    -    Add code to enable In-band ECC feautre option in setup menu.
    +    Add code to enable In-band ECC feature option in setup menu.

    +    Change-Id: I1acdf89ccf5e390d058e6d2a3893c1545282f5fc
         Upstream-Status: Inappropriate [custom feature]
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

 9:  7f67c3d66ef =  8:  d4dc8563016 payloads/external/edk2: Update revision for new memory menu PCDs
10:  27de6e90249 !  9:  a63970dc41f configs/config.hardkernel_odroid_h4: Enable FW performance measurement
    @@ Metadata
     Author: Michał Żygowski <michal.zygowski@3mdeb.com>

      ## Commit message ##
    -    configs/config.hardkernel_odroid_h4: Enable FW performance measurement
    +    configs/config.hardkernel_odroid_h4*: Enable FW performance measurement

         FW performance measurement is required to measure difference in boot
         time with fast boot.

    +    Change-Id: I28071058985f16e4cc1cee936c7a380862e5be64
         Upstream-Status: Inappropriate [custom configs]
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

11:  73500a8ff21 ! 10:  13977990fc2 .github/workflows/build.yml: Build netcard variant of ODROID
    @@ Metadata
      ## Commit message ##
         .github/workflows/build.yml: Build netcard variant of ODROID

    -    Add build checks for new ODROID buidl variant.
    +    Add build checks for new ODROID build variant.

    +    Change-Id: Icf2f3645e295b721356252e996900fd14f278b3b
         Upstream-Status: Inappropriate [CI configuration]
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

12:  b5fe9555b4b ! 11:  258de1fd334 configs/config.hardkernel_odroid_h4: Bump to v0.9.1-rc1
    @@ Metadata
     Author: Michał Żygowski <michal.zygowski@3mdeb.com>

      ## Commit message ##
    -    configs/config.hardkernel_odroid_h4: Bump to v0.9.1-rc1
    +    configs/config.hardkernel_odroid_h4*: Bump to v0.9.1-rc1

    +    Change-Id: I7b695dc1e334bea77480e58a6e1695f440eed0a5
         Upstream-Status: Inappropriate [custom configs]
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

13:  481b2190c68 ! 12:  9d1df050a39 mb/hardkernel/odroid-h4/gpio.c: Fix GPP_A8 macro
    @@ Metadata
      ## Commit message ##
         mb/hardkernel/odroid-h4/gpio.c: Fix GPP_A8 macro

    -    Probably a mistake in upstream implementation.
    -
    +    Change-Id: Id66590bbb4aba3f19d448ebcb08aa544f25b975c
         Upstream-Status: Pending
         Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

14:  43531c3e84f = 13:  a022278daf8 payloads/external/iPXE: Update revision to 2025.03
```

</p>
</details> 

> GPP_A8 macro

That actually came from our version.  Upstream had `PAD_CFG_GPO(GPP_A8, 1, PLTRST),     /* LAN_DISABLE# */`, but we went with our GPIO definitions during latest rebase.
